cvat/apps/dataset_manager/util.py

```diff
@@ -80,3 +92,138 @@ def deepcopy_simple(v):
         return v
     else:
         return deepcopy(v)
+
+
+class LockNotAvailableError(Exception):
+    pass
+
+
+def make_export_lock_key(filename: os.PathLike[str]) -> str:
+    return f"export_lock:{filename}"
+
+@contextmanager
+def get_dataset_cache_lock(
+    export_path: os.PathLike[str],
+    *,
+    ttl: int | timedelta,
+    block: bool = True,
+    acquire_timeout: Optional[int | timedelta] = None,
+) -> Generator[Lock, Any, Any]:
+    if isinstance(acquire_timeout, timedelta):
+        acquire_timeout = acquire_timeout.total_seconds()
+    if acquire_timeout is not None and acquire_timeout < 0:
+        raise ValueError("acquire_timeout must be a non-negative number")
+    elif acquire_timeout is None:
+        acquire_timeout = -1
+
+    if isinstance(ttl, timedelta):
+        ttl = ttl.total_seconds()
+    if not ttl or ttl < 0:
+        raise ValueError("ttl must be a non-negative number")
+
+    # https://redis.io/docs/latest/develop/use/patterns/distributed-locks/
+    # The lock is exclusive, so it may potentially reduce performance in some cases,
+    # where parallel access is potentially possible and valid,
+    # e.g. dataset downloading could use a shared lock instead.
+    lock = Redlock(
+        key=make_export_lock_key(export_path),
+        masters={django_rq.get_connection(settings.CVAT_QUEUES.EXPORT_DATA.value)},
+        auto_release_time=ttl,
+    )
+    try:
+        acquired = lock.acquire(blocking=block, timeout=acquire_timeout)
+        if acquired:
+            yield lock
+        else:
+            raise LockNotAvailableError
+
+    finally:
+        with suppress(ReleaseUnlockedLock):
```

Comment:
Maybe we could at least log a warning if the lock was unlocked? It seems like it could cause hard-to-debugissues, so some diagnostics would be helpful.

Reply:
Updated

Reply:
But there's still no diagnostic.

Reply:
If it happens, it's being logged in both worker callbacks as a regular exception. For requests, it can be logged there once there is such logic.

Reply:
I don't understand, how can it be logged when `get_export_cache_lock` doesn't report this case in any way? Maybe we're talking about different things? I'm talking about the scenario where the code within the scope of the context manager takes too long and the lock unlocks by itself.

Reply:
Ok, restored the exception.